src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,281 @@
+            if (!batchSignersSet.insert(innerAccount).second)
+            {
+                JLOG(ctx.j.trace())
+                    << "BatchTrace[" << batchId
+                    << "]:" << "duplicate signer found: " << innerAccount;
+                return temINVALID_BATCH;
+            }
+        }
+
+        // Check the batch signers signatures.
+        auto const sigResult = ctx.tx.checkBatchSign(
+            STTx::RequireFullyCanonicalSig::yes, ctx.rules);
+
+        if (!sigResult)
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << batchId
+                                << "]:" << "invalid batch txn signature.";
+            return temBAD_SIGNATURE;
+        }
+    }
+
+    std::unordered_set<AccountID> uniqueSigners;
+    std::unordered_set<uint256, beast::uhash<>> uniqueHashes;
+    for (int i = 0; i < txns.size(); ++i)
+    {
+        if (!uniqueHashes.emplace(hashes[i]).second)
+        {
+            JLOG(ctx.j.trace())
+                << "BatchTrace[" << batchId << "]:" << "duplicate TxID found.";
+            return temMALFORMED;
+        }
+
+        STTx const stx = STTx{STObject(txns[i])};
+        if (stx.getTransactionID() != hashes[i])
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << batchId
+                                << "]:" << "txn hash does not match TxIDs hash."
+                                << "index: " << i;
+            return temMALFORMED;
+        }
+
+        if (!stx.isFieldPresent(sfTransactionType))
+        {
+            JLOG(ctx.j.trace())
+                << "BatchTrace[" << batchId
+                << "]:" << "TransactionType missing in inner txn."
+                << "index: " << i;
+            return temINVALID_BATCH;  // LCOV_EXCL_LINE
+        }
+
+        if (stx.getFieldU16(sfTransactionType) == ttBATCH)
+        {
+            JLOG(ctx.j.trace())
+                << "BatchTrace[" << batchId
+                << "]:" << "batch cannot have an inner batch txn."
+                << "index: " << i;
+            return temINVALID_BATCH;
+        }
+
+        if (stx.isFieldPresent(sfTxnSignature))
+        {
+            JLOG(ctx.j.trace())
+                << "BatchTrace[" << batchId
+                << "]:" << "inner txn cannot include TxnSignature."
+                << "index: " << i;
+            return temINVALID_BATCH;
+        }
+
+        if (!stx.getSigningPubKey().empty())
+        {
+            JLOG(ctx.j.trace())
+                << "BatchTrace[" << batchId
+                << "]:" << "inner txn must include empty SigningPubKey."
+                << "index: " << i;
+            return temINVALID_BATCH;
+        }
+
+        if (stx.isFieldPresent(sfSigners))
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << batchId
+                                << "]:" << "inner txn cannot include Signers."
+                                << "index: " << i;
+            return temINVALID_BATCH;
+        }
```

Comment:
These can all be combined into one `if`

Reply:
fixed: https://github.com/XRPLF/rippled/pull/5060/commits/accc8e0f11738205a47cfe619472fdd335bd93ab